ts/repositories.bzl

```diff
@@ -6,24 +6,85 @@ See https://docs.bazel.build/versions/main/skylark/deploying.html#dependencies
+    if rctx.attr.version:
+        version = rctx.attr.version
+    else:
+        p = json.decode(rctx.read(rctx.path(rctx.attr.package_json)))
+        if "devDependencies" in p.keys() and "typescript" in p["devDependencies"]:
+            ts = p["devDependencies"]["typescript"]
+        elif "dependencies" in p.keys() and "typescript" in p["dependencies"]:
+            ts = p["dependencies"]["typescript"]
+        else:
+            fail("key `typescript` not found in either dependencies or devDependencies of %s" % rctx.path(rctx.attr.package_json))
+        if any([not seg.isalnum() for seg in ts.split(".")]):
+            fail("""typescript version in package.json must be exactly specified, not a semver range: %s.
+            You can supply an exact `ts_version` attribute to `rules_ts_dependencies` to bypass this check.""" % ts)
+        version = ts
+
+    if rctx.attr.integrity:
+        integrity = rctx.attr.integrity
+    elif version in TS_VERSIONS.keys():
+        integrity = TS_VERSIONS[version]
+    else:
+        fail("""typescript version {} is not mirrored in rules_ts, is this a real version?
+            If so, you must manually set `ts_integrity`.
+            See documentation on rules_ts_dependencies.""".format(version))
+
+    rctx.download_and_extract(
+        url = [u.format(version) for u in rctx.attr.urls],
+        integrity = integrity,
+    )
+    rctx.symlink(rctx.path(rctx.attr.build_file), "BUILD.bazel")
+
+http_archive_version = repository_rule(
+    implementation = _http_archive_version_impl,
+    attrs = {
+        "integrity": attr.string(doc = "Needed only if the ts version isn't mirrored"),
+        "version": attr.string(doc = "Explicit version for urls placeholder. If provided, the package.json is not read"),
+        "urls": attr.string_list(doc = "URLs to fetch from, each should have a {}-style placeholder"),
+        "build_file": attr.label(doc = "the BUILD file to symlink into the created repository"),
+        "package_json": attr.label(doc = "a path on disk to package.json which may have a version for the package"),
+    },
+)
+
+# WARNING: any additions to this function may be BREAKING CHANGES for users
 # because we'll fetch a dependency which may be different from one that
 # they were previously fetching later in their WORKSPACE setup, and now
 # ours took precedence. Such breakages are challenging for users, so any
 # changes in this function should be marked as BREAKING in the commit message
 # and released only in semver majors.
-def rules_ts_dependencies():
-    "Dependencies needed by users of rules_ts"
+def rules_ts_dependencies(ts_version_from_package_json = None, ts_version = None, ts_integrity = None):
+    """Dependencies needed by users of rules_ts.
+
+    To skip fetching the typescript package, define repository called 'npm_typescript' before calling this.
+
+    Args:
+        ts_version_from_package_json: path to a package.json file which declares an exact typescript version.
```

Comment:
This seems like a weird name to me, with "from"...

Reply:
do you have a different suggestion? I had just `package_json` but then I found myself wanting to comment on the usage site with "This is only used to get the version of typescript"

Reply:
Suggesting a name is a lot harder then complaining about a name 😅 

Anything I can think of has the same issue you mentioned (you need to add a comment saying "all we want is the version").

And just dropping the "from" makes the name more confusing?

I figured it's worth thinking about even if we leave it as is... 🤷

Reply:
it's also more clear this way that `ts_version` and `ts_version_from_package_json` are mutually exclusive?
I added comments about that as well.

Reply:
the other option is only one string-typed attribute `ts_version` with some special value, like

```
ts_version = "4.6.1"
ts_version = "from=//:package_json"
```
but that's inventing syntax and maybe less buildozer-readable

Reply:
Yeah I though of some magic syntax as well but that's not worth it.

> it's also more clear this way that ts_version and ts_version_from_package_json are mutually exclusive?
Those names do make it clearer 👍

Reply:
```
ts_version = "4.6.1"
ts_version = "//:package_json"
```
you could detect if the string is a "version", if not assume it is a label

Reply:
actually, it should be sufficient to check if the first char is a number since labels can't start with numbers

Reply:
you can also run into trouble if the version in package json is non-trivial
- it could be of the form `1.x.x` or `^1` or `~1.2`
- it could be in dependencies or devDependencies
- it could be resolved via "resolutions"
- it could be missing altogether defined as a transitive

Reply:
tho I suppose you'd just error out in those cases

Reply:
> you could detect if the string is a "version", if not assume it is a label

Do we want to avoid string attributes that might be labels though? That always feels odd to me, then bazel can't verify it up front, IDEs don't understand it etc

Reply:
We'll do this in more rulesets like rules_terser etc. so I think it's worth getting what we want.

Advantage of two attributes:
- more machine-editable, you can tell whether a version or a label is specified
- error messages can say "I don't understand your version so please set the xx attribute"

We could make the second attribute name shorter:
```
ts_version = "4.6.1"
ts_version_from = "//:package.json"
```

Reply:
I like the short name, it reads well

Reply:
Suppose if we specify both we fail out?

Reply:
Yes, exactly one of the two is required.